### PR TITLE
feat: rewrite auth

### LINKED ISSUE #365: [Timebox] AAUser, when my token is outdated, it is refreshed and I am not redirected to auth page

### LINKED ISSUE #366: Refresh token shouldn't be called multiple times from FE or it will trigger back-end security

### LINKED ISSUE #367: Tokens shouldn't be store in localStorage

### LINKED ISSUE #368: Some auth controllers can be replaced by resolvers

### PR DESCRIPTION
**Pull request by magrinj**
1. Server

- [x] Move auth controllers to resolvers when possible

2. Front

- [x] Move tokens from local storage to cookie storage, better for security reason (XSS Attack)
- [x] Rewrite Apollo client using an Apollo Factory
- [x] Apollo client now is queuing request during refresh token call to avoid multiple calls to refresh token (if not that is going to trigger the back-end security)
- [x] Add logger and error logging to Apollo in development mode to help improve query calls
Screenshot 2023-06-23 at 2 27 08 PM

- [x] Add listeners system on cookie storage to allow creating hooks to know if user is logged or not
- [x] When refresh-token is expired user is redirect to login page

### CODE REVIEW

File: front/src/modules/utils/cookie-storage.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    this.keys.delete(key);
+    Cookies.remove(key);
+    this.dispatch(key, undefined, oldValue);
+  }
+
+  clear(): void {
+    this.keys.forEach((key) => this.removeItem(key));
+  }
+
+  private dispatch(
```
Comment (lucasbordeau) [inline front/src/modules/utils/cookie-storage.ts:36]: I would try to rely on Recoil for handling state across components as it seems to be feasible and this implementation would introduce a custom event handling system in parallel.

File: front/src/modules/utils/cookie-storage.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import Cookies, { CookieAttributes } from 'js-cookie';
+
+type Listener = (
+  newValue: string | undefined,
+  oldValue: string | undefined,
+) => void;
+
+class CookieStorage {
```
Comment (lucasbordeau) [inline front/src/modules/utils/cookie-storage.ts:8]: Could we try to implement this with hooks ? I don't know how introducing classes will scale in a functional React / Recoil ecosystem.

File: front/src/providers/apollo/ApolloProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { ApolloProvider as ApolloProviderBase } from '@apollo/client';
+import { useRecoilState } from 'recoil';
+
+import { isMockModeState } from '@/auth/states/isMockModeState';
+
+import { apolloClient } from './apollo-client';
+import { mockClient } from './mock-client';
+
+export const ApolloProvider: React.FC<React.PropsWithChildren> = ({
```
Comment (lucasbordeau) [inline front/src/providers/apollo/ApolloProvider.tsx:9]: Could we try to keep the existing pattern which is : function ApolloProvider({ children }: React.PropsWithChildren)

File: front/src/modules/utils/cookie-storage.ts (hunk 1 above)
Comment (magrinj) [reply front/src/modules/utils/cookie-storage.ts:36]: I will take a look on Recoil to see how I can change this

File: front/src/providers/apollo/ApolloProvider.tsx (hunk 3 above)
Comment (magrinj) [reply front/src/providers/apollo/ApolloProvider.tsx:9]: Why not using the official type for functional component, it enforce type and allow the use of `displayName` that can be useful during test.
```typescript
    type FC<P = {}> = FunctionComponent<P>;

    interface FunctionComponent<P = {}> {
        (props: P, context?: any): ReactNode;
        propTypes?: WeakValidationMap<P> | undefined;
        contextTypes?: ValidationMap<any> | undefined;
        defaultProps?: Partial<P> | undefined;
        displayName?: string | undefined;
    }
```

Comment (charlesBochet) [reply front/src/providers/apollo/ApolloProvider.tsx:9]: Let's discuss it, I think consistency is more important overall. If we start using this notation, we should take this decision for the whole repository @magrinj @lucasbordeau

**Review by charlesBochet — APPROVED** [review]
Nice work! Let's rediscuss coding convention together!

**Pull request merged**